cli/azd/pkg/tools/javac/javac.go

```diff
@@ -0,0 +1,124 @@
+package javac
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	osexec "os/exec"
+	"path/filepath"
+
+	"github.com/azure/azure-dev/cli/azd/pkg/exec"
+	"github.com/azure/azure-dev/cli/azd/pkg/tools"
+	"github.com/blang/semver/v4"
+)
+
+const javac = "javac"
+
+type JavacCli interface {
+	tools.ExternalTool
+}
+
+type javacCli struct {
+	cmdRun exec.CommandRunner
+}
+
+func NewCli(cmdRun exec.CommandRunner) JavacCli {
+	return &javacCli{
+		cmdRun: cmdRun,
+	}
+}
+
+func (j *javacCli) VersionInfo() tools.VersionInfo {
+	return tools.VersionInfo{
+		MinimumVersion: semver.Version{
+			Major: 17,
+			Minor: 0,
+			Patch: 0},
+		UpdateCommand: "Visit the website for your installed JDK to upgrade",
+	}
+}
+
+func (j *javacCli) CheckInstalled(ctx context.Context) (bool, error) {
+	path, err := getInstalledPath()
+	if err != nil {
+		return false, err
+	}
+
+	runResult, err := j.cmdRun.Run(ctx, exec.RunArgs{
+		Cmd:  path,
+		Args: []string{"--version"},
+	})
+	if err != nil {
+		return false, fmt.Errorf("checking javac version: %w", err)
+	}
+
+	jdkVer, err := tools.ExtractSemver(runResult.Stdout)
+	if err != nil {
+		return false, fmt.Errorf("converting to semver version fails: %w", err)
+	}
+
+	requiredVersion := j.VersionInfo()
+	if jdkVer.LT(requiredVersion.MinimumVersion) {
+		return false, &tools.ErrSemver{ToolName: j.Name(), VersionInfo: requiredVersion}
+	}
+
+	return true, nil
+}
+
+func (j *javacCli) InstallUrl() string {
+	return "https://www.microsoft.com/openjdk"
+}
+
+func (j *javacCli) Name() string {
+	return "Java JDK"
+}
+
+// getInstalledPath returns the installed javac path.
+//
+// javac is located by consulting, in search order:
+//   - JAVA_HOME (if set)
+//   - PATH
+//
+// An error is returned if javac could not be found, or if invalid locations are provided.
+func getInstalledPath() (string, error) {
+	path, err := findByEnvVar("JAVA_HOME")
+	if path != "" {
+		return path, nil
+	}
+	if err != nil {
+		return "", fmt.Errorf("JAVA_HOME is set to an invalid directory: %w", err)
+	}
+
+	path, err = osexec.LookPath(javac)
+	if err == nil {
+		return path, nil
+	}
+
+	if !errors.Is(err, osexec.ErrNotFound) {
+		return "", fmt.Errorf("failed looking up javac in PATH: %w", err)
+	}
+
+	return "", errors.New(
+		"javac could not be found. Set JAVA_HOME environment variable to point to your Java JDK installation, " +
+			"or include javac in your PATH environment variable")
+}
+
+// findByEnvVar returns the javac path by the following environment variable home directory.
+//
+// An error is returned if an error occurred while finding.
+// If the environment variable home directory is unset, an empty string is returned with no error.
+func findByEnvVar(envVar string) (string, error) {
+	home := os.Getenv(envVar)
```

Comment:
Folks can also do `azd env set JAVA_HOME` to set the java path only for `azd`.

We need to build the general solution in environment package for getting values from system environment, by first looking into .env from azd and latter looking at system environment.  While this is not yet implemented, each component is currently checking both .env and system environment

Reply:
I'm not completely sold on that idea, and I don't think that's been the intended usage based on what I've seen, but I could be wrong.

The environment file is currently used for:
1. Storing post-deployment, infrastructure provider state in a local file
2. Storing app settings configuration (the runtime application uses the configuration file)

I'm not sure if we further increase the footprint to "build systems and languages". The value of having it stored inside `.env` file to form an isolation seems like a poor fit from an actual solution (docker containers).

Reply:
That's how I was looking at it as well at some point. Just as a state-persist strategy.
But I had a conversation with @ellismg and learned that it is not just a way to record settings, but all those settings are injected as part of user's environment and overrides any system environment when running external tools (like npm for deploy)

Reply:
In a few places in the code we've used this pattern of first considering the values in the environment and falling back to the os system environment variables (https://github.com/Azure/azure-dev/blob/8bb8e2d992b7fb2878d2be489adaedac5d567e93/cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go#L516-L520) if we didn't find something.

We also have issues around azd loading `.env` files from other locations, not just `.azd\<environment-name>\.env`, with the thinking that you could have a root `.env` file and then override values with environment specific files `.env` files.  It feels like that's a sensible long-term design, and the story were can tell is "like lots of other tools, you can use `.env` files to provide an additional set of environment variables which overwrite system values and `azd` ensures these files get loaded on your behalf.

In that idealized world, looking at the .env file first and then falling back to the system makes sense.

But, tactically, we are pretty inconsistent here and I don't think we have a complete design.

Given that JAVA_HOME is unlikely to vary across environments - I think I'm fine with just calling `os.Getenv` right now.  If we make changes here, we'll have to audit all the calls to `os.Getenv` any way to understand their intention, so I'm not worried about us going back and fixing this later - it will be easy to spot.

I think as we move away from using a `.env` file to store all configuration state and towards something more structured, in a different file, we can move to a world where it's easier for us to say exactly what a `.env` file holds and when to load from it.